kevm_pyk/solc_to_k.py

```diff
@@ -0,0 +1,289 @@
+import functools
+
+from pyk import *
+from pyk.ktool import KPrint
+
+from .utils import intersperse
+
+
+def solc_to_k(*, command: str, kompiled_directory: str, contract_file: str, contract_name: str, solc_output):
+    assert command == 'solc-to-k'
+    kevm = KPrint(kompiled_directory)
+    kevm.symbolTable = kevmSymbolTable(kevm.symbolTable)
+
+    contract_json = json.loads(solc_output.read())['contracts'][contract_file + ':' + contract_name]
+    storage_layout = contract_json['storage-layout']
+    abi = contract_json['abi']
+    bin_runtime = '0x' + contract_json['bin-runtime']
+
+    # TODO: add check to kevm:
+    # solc version should be >=0.8.0 due to:
+    # https://github.com/ethereum/solidity/issues/10276
+
+    contract_sort = KSort(f'{contract_name}Contract')
+
+    storage_sort = KSort(f'{contract_name}Storage')
+    storage_sentences = _extract_storage_sentences(contract_name, storage_sort, storage_layout)
+    storage_productions, storage_rules = map(list, zip(*storage_sentences))
+    storage_location_production = KProduction([KNonTerminal(contract_sort), KTerminal('.'), KNonTerminal(storage_sort)], KSort('Int'), att=KAtt({'klabel': f'storage_{contract_name}', 'macro': ''}))
+
+    function_sort = KSort(f'{contract_name}Function')
+    function_sentences = _extract_function_sentences(contract_name, function_sort, abi)
+    function_productions, function_rules = map(list, zip(*function_sentences))
+    function_call_data_production = KProduction([KNonTerminal(contract_sort), KTerminal('.'), KNonTerminal(function_sort)], KSort('ByteArray'), att=KAtt({'klabel': f'function_{contract_name}', 'macro': ''}))
+
+    unknownBoolProduction = KProduction([KTerminal('unknownBool')], KSort('Bool'), att=KAtt({'klabel': 'unknownBool', 'symbol': '', 'no-evaluators': '', 'function': '', 'functional': ''}))
+
+    binRuntimeProduction = KProduction([KTerminal('#binRuntime'), KTerminal('('), KNonTerminal(contract_sort), KTerminal(')')], KSort('ByteArray'), att=KAtt({'klabel': 'binRuntime', 'macro': ''}))
+
+    contractProduction = KProduction([KTerminal(contract_name)], contract_sort, att=KAtt({'klabel': f'contract_{contract_name}'}))
+    contractMacro = KRule(KRewrite(KApply('binRuntime', [KConstant(contract_name)]), _parseByteStack(_stringToken(bin_runtime))))
+
+    binRuntimeModuleName = contract_name.upper() + '-BIN-RUNTIME'
+    binRuntimeModule = KFlatModule(
+            binRuntimeModuleName,
+            [KImport('EDSL', True)],
+            [unknownBoolProduction, contractProduction] +
+            storage_productions +
+            [storage_location_production] +
+            storage_rules +
+            function_productions +
+            [function_call_data_production] +
+            function_rules +
+            [binRuntimeProduction, contractMacro]
+        )
```

Comment:
```suggestion
    binRuntimeModule = KFlatModule( binRuntimeModuleName
                                  , [KImport('EDSL', True)]
                                  ,   [unknownBoolProduction, contractProduction]
                                    + storage_productions
                                    + [storage_location_production]
                                    + storage_rules
                                    + function_productions
                                    + [function_call_data_production]
                                    + function_rules
                                    + [binRuntimeProduction, contractMacro]
                                  )
```

In general I prefer the join operator for associative lists to show up at the beginning of a line, not at the end, so it's easier to see where it starts/ends. May require putting `\` at the end of the line too.

I also like, as much as possible, for things to line up horizontally (eg. below where I line up the `=` sign in the symbolTable definition). This makes it much easier to scan the structure of the file based on its "shape".

Reply:
In this case, the `_,_` and the `_+_` are associative join operators.